File: src/com/libill/demos/activity/UDPActivity.java
package com.libill.demos.activity;

import android.app.Activity;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;

import com.libill.demos.R;

public class UDPActivity extends Activity {
    private EditText et_ip;
    private EditText et_content;
    private Button bt_send_data;
    DatagramSocket socket = null;
    InetAddress serverAddress = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_udp);

        et_ip = (EditText) findViewById(R.id.et_ip);
        et_content = (EditText) findViewById(R.id.et_content);
        bt_send_data = (Button) findViewById(R.id.bt_send);


        bt_send_data.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View arg0) {
                if(socket == null) {
                    try {
                        socket = new DatagramSocket(8888);
                        serverAddress = InetAddress.getByName(et_ip.getText().toString());
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            String sendData = et_content.getText().toString();
                            byte data[] = sendData.getBytes();
                            DatagramPacket packet = new DatagramPacket(data, data.length, serverAddress, 8888);
                            socket.send(packet);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }).start();
            }
        });
    }
}

File: app/src/main/java/com/libill/demos/activity/UDPActivity.kt
package com.libill.demos.activity

import android.os.Bundle
import com.libill.demos.base.BaseActivity
import com.libill.demos.databinding.ActivityUdpBinding
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

class UDPActivity : BaseActivity() {
    private var socket: DatagramSocket? = null
    private var serverAddress: InetAddress? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ActivityUdpBinding.inflate(layoutInflater).apply {
            setContentView(this.root)
            btSend.setOnClickListener {
                if (socket == null) {
                    try {
                        socket = DatagramSocket(8888)
                        serverAddress = InetAddress.getByName(etIp.text.toString())
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
                Thread {
                    try {
                        val sendData = etContent.text.toString()
                        val data = sendData.toByteArray()
                        val packet = DatagramPacket(data, data.size, serverAddress, 8888)
                        socket!!.send(packet)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }.start()
            }
        }
    }
}